test/e2e/azure_vmextensions.go

```diff
@@ -0,0 +1,120 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-11-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AzureVMExtensionsSpecInput is the input for AzureVMExtensionsSpec.
+type AzureVMExtensionsSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	VMName                string
+	ClusterName           string
+	SkipCleanup           bool
+}
+
+// AzureVMExtensionsSpec implements a test that verifies VM extensions are created and deleted.
+func AzureVMExtensionsSpec(ctx context.Context, inputGetter func() AzureVMExtensionsSpecInput) {
+	var (
+		specName = "azure-vmextensions"
+		input    AzureVMExtensionsSpecInput
+	)
+
+	Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+	Expect(input.Namespace).ToNot(BeNil(), "Invalid argument. input.Namespace can't be nil when calling %s spec", specName)
+	Expect(input.ClusterName).ToNot(BeEmpty(), "Invalid argument. input.ClusterName can't be empty when calling %s spec", specName)
+
+	By("creating a Kubernetes client to the workload cluster")
+	workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, input.Namespace.Name, input.ClusterName)
+	Expect(workloadClusterProxy).NotTo(BeNil())
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	By("Retrieving all machines from the machine template spec")
+	machineList := &infrav1.AzureMachineList{}
+	// list all of the requested objects within the cluster namespace with the cluster name label
+	Logf("Listing machines in namespace %s with label %s=%s", input.Namespace.Name, clusterv1.ClusterLabelName, workloadClusterProxy.GetName())
+	err := mgmtClient.List(ctx, machineList, client.InNamespace(input.Namespace.Name), client.MatchingLabels{clusterv1.ClusterLabelName: workloadClusterProxy.GetName()})
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Ensuring the machine spec contains expected VM extensions")
+	for _, machine := range machineList.Items {
+		if strings.HasPrefix(machine.Name, input.VMName) {
+			Expect(len(machine.Spec.VMExtensions)).To(Equal(1))
+			Expect(machine.Spec.VMExtensions[0].Name).To(Equal("CustomScript"))
```

Comment:
so here our test is essentially validating the yaml template and expecting a very precise extension to be present on a very specific machine at a hardcoded `0` position. This could cause issues in the future if we add extensions, rename the machine, etc.

Since the main goal of the e2e test is to test CAPZ code (in this case that adding a VM extension in the AzureMachine Spec results in the Azure VM that gets created to have that extension), what would you think of rewriting the test to be a bit more generic, ie. For all extensions defined in all the AzureMachines in the cluster, verify that the extension exists in the corresponding VM in Azure. In otherwords, we could iterate through the machines like you're doing here, and then check that the VM for each machine has whatever extensions defined in spec.

That way, this test could also be reused for multiple cluster templates that have different extensions and AzureMachine configurations (maybe even several in AzureMachines in one cluster) and wouldn't need to be enabled just for the specific cluster test / template combination that has the extension in the template.

Reply:
This is a much better approach! I'm trying to implement this now and I'm having some trouble trying to get the machine spec, since the windows machine names on Azure are different than the machine names defined in the spec. Is there a way to get a Machine other than the name itself?

Reply:
I think my idea right now is to create a mapping of ProviderIDs to the extension names from the spec, then iterate through the machines and verify the extensions using the ID as a key. I'm assuming that the spec's ProviderID is equal to the Virtual Machine's ID

Reply:
> I'm assuming that the spec's ProviderID is equal to the Virtual Machine's ID

that's correct

Reply:
I did this by creating a map of machineIDs to an array of expected VM extensions from the spec, and ensured that the VM extensions from Azure matched it. I'm not sure if this is the best way to go about it, but let me know what you think! @CecileRobertMichon 

[Code](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2631/files#diff-2bea7aae68fda3b77b9d0180bd026269d0804ded551d4d9d5897e528cf20da31R72-R78)